File: raw/github/paypal/paypal-ios-sdk-demo-app/snapshots/2026-08-16-047a50e/files/PayPalDemo/ViewModels/CardCheckoutValidationViewModel.swift
import Foundation
import CardPayments

class CardCheckoutValidationViewModel: ObservableObject {
    @Published var errorMessage: String = ""
    
    @Published var cardNumber: String = "4111 1111 1111 1111"
    @Published var expirationDate: String = "01 / 27"
    @Published var cvv: String = "123"

    private let cardFormatter = CardFormatter()

    var isValid: Bool {
        return Card.isCardFormValid(cardNumber: cardNumber, expirationDate: expirationDate, cvv: cvv)
    }

    func isCardValid(completion: (Card?) -> Void) {
        if isValid {
            let card = Card.createCard(cardNumber: cardNumber, expirationDate: expirationDate, cvv: cvv)
            completion(card)
        } else {
            errorMessage = "Invalid card details. Please check and try again."
            completion(nil)
        }
    }

    func updateCardNumber(_ newValue: String) {
        cardNumber = cardFormatter.formatFieldWith(cardNumber, field: .cardNumber)
        cvv = CardType.unknown.getCardType(cardNumber) == .americanExpress ? "1234" : "123"
    }

    func updateExpirationDate(_ newValue: String) {
        expirationDate = cardFormatter.formatFieldWith(newValue, field: .expirationDate)
    }

    func updateCVV(_ newValue: String) {
        cvv = cardFormatter.formatFieldWith(cvv, field: .cvv)
    }
}

extension Card {
    static func createCard(cardNumber: String, expirationDate: String, cvv: String) -> Card {
        let cleanedCardText = cardNumber.replacingOccurrences(of: " ", with: "")

        let expirationComponents = expirationDate.components(separatedBy: " / ")
        let expirationMonth = expirationComponents[0]
        let expirationYear = "20" + expirationComponents[1]

        return Card(number: cleanedCardText, expirationMonth: expirationMonth, expirationYear: expirationYear, securityCode: cvv)
    }

    static func isCardFormValid(cardNumber: String, expirationDate: String, cvv: String) -> Bool {
        let cleanedCardNumber = cardNumber.replacingOccurrences(of: " ", with: "")
        let cleanedExpirationDate = expirationDate.replacingOccurrences(of: " / ", with: "")

        let enabled = cleanedCardNumber.count >= 15 && cleanedCardNumber.count <= 19
        && cleanedExpirationDate.count == 4 && cvv.count >= 3 && cvv.count <= 4
        return enabled
    }
}

File: raw/github/paypal/paypal-android-sdk-demo-app/snapshots/2026-08-16-d1137d5/files/app/src/main/java/com/firstapp/paypaldemo/cardcheckout/CardCheckoutValidationViewModel.kt
package com.firstapp.paypaldemo.cardcheckout

import androidx.lifecycle.ViewModel
import com.paypal.android.cardpayments.Card

class CardCheckoutValidationViewModel : ViewModel() {

    var errorMessage: String = ""
    var cardNumber: String = "4111111111111111"
    var expirationDate: String = "0127"
    var cvv: String = "123"

    private val cardFormatter = CardFormatter()

    val isValid: Boolean
        get() = isCardFormValid()

    /**
     * Check basic length constraints:
     * - Card # must be 15-19 digits
     * - Expiration date must be exactly 4 digits (MMYY)
     * - CVV must be 3 or 4 digits
     */
    private fun isCardFormValid(): Boolean {
        val cleanedNumber = cardNumber
        val cleanedExp = expirationDate
        val cleanedCvv = cvv

        val numberOk = cleanedNumber.length in 15..19
        val expOk = (cleanedExp.length == 4)
        val cvvOk = (cleanedCvv.length in 3..4)

        return numberOk && expOk && cvvOk
    }

    /**
     * If valid, return a Card object we can send to `CardPaymentViewModel`.
     * Otherwise set errorMessage and return null.
     */
    fun validCard(): Card? {
        return if (isValid) {
            // create a Card object
            val cleanedNumber = cardNumber
            val exp = expirationDate
            val month = exp.take(2)
            val year = "20" + exp.drop(2)
            Card(
                number = cleanedNumber,
                expirationMonth = month,
                expirationYear = year,
                securityCode = cvv
            )
        } else {
            errorMessage = "Invalid card details. Please check and try again."
            null
        }
    }

    fun updateCardNumber(newValue: String) {
        cardNumber = cardFormatter.formatFieldWith(newValue, CardFields.CARD_NUMBER)
        //  "cvv = 1234 if AMEX else 123"
        val type = CardType.getCardType(cardNumber)
        if (type == CardType.AMERICAN_EXPRESS && cvv.length < 4) {
            cvv = "1234"
        } else if (type != CardType.AMERICAN_EXPRESS && cvv.length > 3) {
            cvv = "123"
        }
    }

    fun updateExpirationDate(newValue: String) {
        val cleaned = newValue.filter { it.isDigit() }.take(4)
        expirationDate = cleaned
    }

    fun updateCVV(newValue: String) {
        cvv = cardFormatter.formatFieldWith(newValue, CardFields.CVV)
    }
}